arch/src/aarch64/fdt.rs

```diff
@@ -0,0 +1,610 @@
+// Copyright 2020 Arm Limited (or its affiliates). All rights reserved.
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the THIRD-PARTY file.
+
+use libc::{c_char, c_int, c_void};
+use std::collections::HashMap;
+use std::ffi::{CStr, CString, NulError};
+use std::fmt::Debug;
+use std::ptr::null;
+use std::{io, result};
+
+use super::super::DeviceType;
+use super::super::InitrdConfig;
+use super::get_fdt_addr;
+use super::gic::GICDevice;
+use super::layout::FDT_MAX_SIZE;
+use crate::aarch64::fdt::Error::CstringFDTTransform;
+use vm_memory::{Address, Bytes, GuestAddress, GuestMemory, GuestMemoryError, GuestMemoryMmap};
+
+// This is a value for uniquely identifying the FDT node declaring the interrupt controller.
+const GIC_PHANDLE: u32 = 1;
+// This is a value for uniquely identifying the FDT node containing the clock definition.
+const CLOCK_PHANDLE: u32 = 2;
+// Read the documentation specified when appending the root node to the FDT.
+const ADDRESS_CELLS: u32 = 0x2;
+const SIZE_CELLS: u32 = 0x2;
+
+// As per kvm tool and
+// https://www.kernel.org/doc/Documentation/devicetree/bindings/interrupt-controller/arm%2Cgic.txt
+// Look for "The 1st cell..."
+const GIC_FDT_IRQ_TYPE_SPI: u32 = 0;
+const GIC_FDT_IRQ_TYPE_PPI: u32 = 1;
+
+// From https://elixir.bootlin.com/linux/v4.9.62/source/include/dt-bindings/interrupt-controller/irq.h#L17
+const IRQ_TYPE_EDGE_RISING: u32 = 1;
+const IRQ_TYPE_LEVEL_HI: u32 = 4;
+
+// This links to libfdt which handles the creation of the binary blob
+// flattened device tree (fdt) that is passed to the kernel and indicates
+// the hardware configuration of the machine.
+#[link(name = "fdt")]
+extern "C" {
+    fn fdt_create(buf: *mut c_void, bufsize: c_int) -> c_int;
+    fn fdt_finish_reservemap(fdt: *mut c_void) -> c_int;
+    fn fdt_begin_node(fdt: *mut c_void, name: *const c_char) -> c_int;
+    fn fdt_property(fdt: *mut c_void, name: *const c_char, val: *const c_void, len: c_int)
+        -> c_int;
+    fn fdt_end_node(fdt: *mut c_void) -> c_int;
+    fn fdt_open_into(fdt: *const c_void, buf: *mut c_void, bufsize: c_int) -> c_int;
+    fn fdt_finish(fdt: *const c_void) -> c_int;
+    fn fdt_pack(fdt: *mut c_void) -> c_int;
+}
```

Comment:
Oh that's not great we have to rely on some external C library to generate the device tree blob. Any chance there's a Rust native way of doing this?
Also, have you thought about using ACPI instead of DeviceTree? I'm asking because it's already part of Cloud-Hypervisor and that would be nice to reuse this part between `x86_64` and `aarch64`.

Reply:
Thanks @sboeuf for pointing these out. I actually think the idea of replacing the fdt implementation with Rust native code makes a lot of sense. However I do have a few concerns and would like to discuss them with you:

**Speaking of the FDT:**
- We did think of rewrote everything completely in Rust at the beginning. The reasons why we used the fdt in current way are that there is an existing `libfdt-dev` which contains everything we need, and rewriting everything might take longer than expected (as we probably need to take care of everything in the library). Existing rust crates related to the DeviceTree (for example: [device_tree](https://docs.rs/device_tree/1.1.0/device_tree/), [dtb](https://github.com/hermitcore/dtb) and [dts_viewer](https://github.com/yodaldevoid/dts_viewer)) are crates that either parsing dtb to dts or simply parsing dts files without compiling them to dtb. All of them cannot satisfy the requirements needed for cloud-hypervisor project.

- We do have an existing discussion about rewriting fdt in the rust-vmm community about this: [vmm-fdt](https://github.com/rust-vmm/community/issues/87), please kindly leave some comments about this and these comments are greatly appreciated.

- As currently it seems that this PR is blocking further PRs from @michael2012z, if the community is not happy with current fdt implementation, I will remove the commit related to the fdt and we can merge the other modules first. Would this be ok with you? Thanks!

**Speaking of the ACPI:**
Yes of course we are going to support ACPI for arm64 (I think crosVM has already done that?...not sure...). Since ACPI also contains a lot of things, currently for the first step of running cloud-hypervisor on arm we just used the DeviceTree as the it is simpler. The supporting of ACPI on arm64 will be submitted as further PRs after we prove the cloud-hypervisor works well on arm64. Hope it is ok :))

Reply:
>Speaking of the FDT

Ok I'm glad to see your actively trying to solve this with a Rust implementation, so I'm okay with using `libfdt-dev` as a temporary solution. I think extending one of the existing crate mentioned above would make more sense since the generation of a fdt blob is not specific to VMMs or virtualization.

> Speaking of the ACPI

Also glad to see you're considering ACPI support, but I understand you need to validate you can make things work with some already existing code before you can go further and refactor/redesign the code.

@sameo @rbradford WDYT?

Reply:
If there is momentum behind creating a Rust native version for FDT manipulation that is good. As you can see we did the same thing for creating ACPI tables.

As for ACPI boot with ARM I checked the Linux source code and it only works with either getting it from EFI or by scanning the memory. The former will limit the ability to do direct kernel boot but we do locate the ACPI tables within the scannable location.